#DataFrames: Using Julia

## DataFrames package is required to read any data frame

Install DataFrames Package in Julia

In [1]:
Pkg.add("DataFrames")


INFO: Installing ArrayViews v0.6.4
INFO: Installing DataArrays v0.2.20
INFO: Installing DataFrames v0.6.10
INFO: Installing Docile v0.5.19
INFO: Installing GZip v0.2.18
INFO: Installing Reexport v0.0.3
INFO: Installing SortingAlgorithms v0.0.6
INFO: Installing StatsBase v0.7.4
INFO: Installing StatsFuns v0.2.0
INFO: Package database updated


Declear DataFrames package

In [2]:
using DataFrames

Set Default display size = 10 in IJulia Notebook

In [3]:
ENV["LINES"] =10

10

##Read data from csv file

In [4]:
df = readtable("home_data.csv")

21613x21 DataFrame
| Row   | id         | date              | price  | bedrooms | bathrooms |
|-------|------------|-------------------|--------|----------|-----------|
| 1     | 7129300520 | "20141013T000000" | 221900 | 3        | 1.0       |
⋮
| 21612 | 291310100  | "20150116T000000" | 400000 | 3        | 2.5       |
| 21613 | 1523300157 | "20141015T000000" | 325000 | 2        | 0.75      |

| Row   | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-------|-------------|----------|--------|------------|------|-----------|
| 1     | 1180        | 5650     | 1.0    | 0          | 0    | 3         |
⋮
| 21612 | 1600        | 2388     | 2.0    | 0          | 0    | 3         |
| 21613 | 1020        | 1076     | 2.0    | 0          | 0    | 3         |

| Row   | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|-------|-------|------------|---------------|----------|--------------|
| 1     | 7     | 1180       | 0             | 1955     | 0            |
⋮
| 21612 | 8     | 1600       | 0             | 2004     | 0            |
| 21613 | 7     | 1020       | 0             | 2008     | 0            |

| Row   | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|-------|---------|---------|----------|---------------|------------|
| 1     | 98178   | 47.5112 | -122.257 | 1340          | 5650       |
⋮
| 21612 | 98027   | 47.5345 | -122.069 | 1410          | 1287       |
| 21613 | 98144   | 47.5941 | -122.299 | 1020          | 1357       |

### Print the column names

In [20]:
 a = names(df);
println(" Column names : ", a)

 Column names : [:id,:date,:price,:bedrooms,:bathrooms,:sqft_living,:sqft_lot,:floors,:waterfront,:view,:condition,:grade,:sqft_above,:sqft_basement,:yr_built,:yr_renovated,:zipcode,:lat,:long,:sqft_living15,:sqft_lot15]


###Print top 6 rows of the DataFrame

In [5]:
head(df)

6x21 DataFrame
| Row | id         | date              | price   | bedrooms | bathrooms |
|-----|------------|-------------------|---------|----------|-----------|
| 1   | 7129300520 | "20141013T000000" | 221900  | 3        | 1.0       |
⋮
| 5   | 1954400510 | "20150218T000000" | 510000  | 3        | 2.0       |
| 6   | 7237550310 | "20140512T000000" | 1225000 | 4        | 4.5       |

| Row | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-----|-------------|----------|--------|------------|------|-----------|
| 1   | 1180        | 5650     | 1.0    | 0          | 0    | 3         |
⋮
| 5   | 1680        | 8080     | 1.0    | 0          | 0    | 3         |
| 6   | 5420        | 101930   | 1.0    | 0          | 0    | 3         |

| Row | grade | sqft_above | sqft_basement | yr_built | yr_renovated | zipcode |
|-----|-------|------------|---------------|----------|--------------|---------|
| 1   | 7     | 1180       | 0             | 1955     | 0            | 98178   |
⋮
| 5   | 8     | 1680       | 0             | 1987     | 0            | 98074   |
| 6   | 11    | 3890       | 1530          | 2001     | 0            | 98053   |

| Row | lat     | long     | sqft_living15 | sqft_lot15 |
|-----|---------|----------|---------------|------------|
| 1   | 47.5112 | -122.257 | 1340          | 5650       |
⋮
| 5   | 47.6168 | -122.045 | 1800          | 7503       |
| 6   | 47.6561 | -122.005 | 4760          | 101930     |

###Print bottom 6 rows of the DataFrame

In [6]:
tail(df)

6x21 DataFrame
| Row | id         | date              | price  | bedrooms | bathrooms |
|-----|------------|-------------------|--------|----------|-----------|
| 1   | 2997800021 | "20150219T000000" | 475000 | 3        | 2.5       |
⋮
| 5   | 291310100  | "20150116T000000" | 400000 | 3        | 2.5       |
| 6   | 1523300157 | "20141015T000000" | 325000 | 2        | 0.75      |

| Row | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-----|-------------|----------|--------|------------|------|-----------|
| 1   | 1310        | 1294     | 2.0    | 0          | 0    | 3         |
⋮
| 5   | 1600        | 2388     | 2.0    | 0          | 0    | 3         |
| 6   | 1020        | 1076     | 2.0    | 0          | 0    | 3         |

| Row | grade | sqft_above | sqft_basement | yr_built | yr_renovated | zipcode |
|-----|-------|------------|---------------|----------|--------------|---------|
| 1   | 8     | 1180       | 130           | 2008     | 0            | 98116   |
⋮
| 5   | 8     | 1600       | 0             | 2004     | 0            | 98027   |
| 6   | 7     | 1020       | 0             | 2008     | 0            | 98144   |

| Row | lat     | long     | sqft_living15 | sqft_lot15 |
|-----|---------|----------|---------------|------------|
| 1   | 47.5773 | -122.409 | 1330          | 1265       |
⋮
| 5   | 47.5345 | -122.069 | 1410          | 1287       |
| 6   | 47.5941 | -122.299 | 1020          | 1357       |

###Find the dimension of DataFrame

In [7]:
size(df)

(21613,21)

size prints DataFrame  df has 21613 rows and 21 columns

###Summarize the total number of bedrooms in the dataset

In [8]:
describe(df[:bedrooms])

Summary Stats:
Mean:         3.370842
Minimum:      0.000000
1st Quartile: 3.000000
Median:       3.000000
3rd Quartile: 4.000000
Maximum:      33.000000


###Find the mean value of price in the dataset

In [9]:
mean(df[:price])

540088.1419053348

###Find the median value of sqft_living

In [10]:
median(df[:sqft_living])

1910.0

### Remove id 

In [21]:
df2 = delete!(df,:id);
head(df2)

6x20 DataFrame
| Row | date              | price   | bedrooms | bathrooms | sqft_living |
|-----|-------------------|---------|----------|-----------|-------------|
| 1   | "20141013T000000" | 221900  | 3        | 1.0       | 1180        |
⋮
| 5   | "20150218T000000" | 510000  | 3        | 2.0       | 1680        |
| 6   | "20140512T000000" | 1225000 | 4        | 4.5       | 5420        |

| Row | sqft_lot | floors | waterfront | view | condition | grade | sqft_above |
|-----|----------|--------|------------|------|-----------|-------|------------|
| 1   | 5650     | 1.0    | 0          | 0    | 3         | 7     | 1180       |
⋮
| 5   | 8080     | 1.0    | 0          | 0    | 3         | 8     | 1680       |
| 6   | 101930   | 1.0    | 0          | 0    | 3         | 11    | 3890       |

| Row | sqft_basement | yr_built | yr_renovated | zipcode | lat     | long     |
|-----|---------------|----------|--------------|---------|---------|----------|
| 1   | 0             | 1955     | 0            | 98178   | 47.5112 | -122.257 |
⋮
| 5   | 0             | 1987     | 0            | 98074   | 47.6168 | -122.045 |
| 6   | 1530          | 2001     | 0            | 98053   | 47.6561 | -122.005 |

| Row | sqft_living15 | sqft_lot15 |
|-----|---------------|------------|
| 1   | 1340          | 5650       |
⋮
| 5   | 1800          | 7503       |
| 6   | 4760          | 101930     |

###Subset data for house with 2 bedrooms

In [11]:
twoBedroom_data = df[df[:bedrooms] .== 2,:]

2760x21 DataFrame
| Row  | id         | date              | price  | bedrooms | bathrooms |
|------|------------|-------------------|--------|----------|-----------|
| 1    | 5631500400 | "20150225T000000" | 180000 | 2        | 1.0       |
⋮
| 2759 | 1523300141 | "20140623T000000" | 402101 | 2        | 0.75      |
| 2760 | 1523300157 | "20141015T000000" | 325000 | 2        | 0.75      |

| Row  | sqft_living | sqft_lot | floors | waterfront | view | condition |
|------|-------------|----------|--------|------------|------|-----------|
| 1    | 770         | 10000    | 1.0    | 0          | 0    | 3         |
⋮
| 2759 | 1020        | 1350     | 2.0    | 0          | 0    | 3         |
| 2760 | 1020        | 1076     | 2.0    | 0          | 0    | 3         |

| Row  | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|------|-------|------------|---------------|----------|--------------|
| 1    | 6     | 770        | 0             | 1933     | 0            |
⋮
| 2759 | 7     | 1020       | 0             | 2009     | 0            |
| 2760 | 7     | 1020       | 0             | 2008     | 0            |

| Row  | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|------|---------|---------|----------|---------------|------------|
| 1    | 98028   | 47.7379 | -122.233 | 2720          | 8062       |
⋮
| 2759 | 98144   | 47.5944 | -122.299 | 1020          | 2007       |
| 2760 | 98144   | 47.5941 | -122.299 | 1020          | 1357       |

Important point to note here is julia requires elementwise operations for subsetting "." [dot] is required for elementwise operations.

In [12]:
size(twoBedroom_data)

(2760,21)

There are 2760 houses with 2 bedrooms

###Subset the data for house with 2 bedrooms and sqft_living more than 1000

In [13]:
twoBedrooms_1000_sqft = df[(df[:bedrooms].== 2).*(df[:sqft_living] .>= 1000),:]

1824x21 DataFrame
| Row  | id         | date              | price  | bedrooms | bathrooms |
|------|------------|-------------------|--------|----------|-----------|
| 1    | 9212900260 | "20140527T000000" | 468000 | 2        | 1.0       |
⋮
| 1823 | 1523300141 | "20140623T000000" | 402101 | 2        | 0.75      |
| 1824 | 1523300157 | "20141015T000000" | 325000 | 2        | 0.75      |

| Row  | sqft_living | sqft_lot | floors | waterfront | view | condition |
|------|-------------|----------|--------|------------|------|-----------|
| 1    | 1160        | 6000     | 1.0    | 0          | 0    | 4         |
⋮
| 1823 | 1020        | 1350     | 2.0    | 0          | 0    | 3         |
| 1824 | 1020        | 1076     | 2.0    | 0          | 0    | 3         |

| Row  | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|------|-------|------------|---------------|----------|--------------|
| 1    | 7     | 860        | 300           | 1942     | 0            |
⋮
| 1823 | 7     | 1020       | 0             | 2009     | 0            |
| 1824 | 7     | 1020       | 0             | 2008     | 0            |

| Row  | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|------|---------|---------|----------|---------------|------------|
| 1    | 98115   | 47.69   | -122.292 | 1330          | 6000       |
⋮
| 1823 | 98144   | 47.5944 | -122.299 | 1020          | 2007       |
| 1824 | 98144   | 47.5941 | -122.299 | 1020          | 1357       |

In [14]:
size(twoBedrooms_1000_sqft)

(1824,21)

There are 1824 houses with 2 bedrooms with sqft_living more than 1000 sqft

### Sort the dataset according to the increasing order of price 

In [15]:
sorted_data = sort(df,cols = :price)

21613x21 DataFrame
| Row   | id         | date              | price   | bedrooms | bathrooms |
|-------|------------|-------------------|---------|----------|-----------|
| 1     | 3421079032 | "20150217T000000" | 75000   | 1        | 0.0       |
⋮
| 21612 | 9808700762 | "20140611T000000" | 7062500 | 5        | 4.5       |
| 21613 | 6762700020 | "20141013T000000" | 7700000 | 6        | 8.0       |

| Row   | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-------|-------------|----------|--------|------------|------|-----------|
| 1     | 670         | 43377    | 1.0    | 0          | 0    | 3         |
⋮
| 21612 | 10040       | 37325    | 2.0    | 1          | 2    | 3         |
| 21613 | 12050       | 27600    | 2.5    | 0          | 3    | 4         |

| Row   | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|-------|-------|------------|---------------|----------|--------------|
| 1     | 3     | 670        | 0             | 1966     | 0            |
⋮
| 21612 | 11    | 7680       | 2360          | 1940     | 2001         |
| 21613 | 13    | 8570       | 3480          | 1910     | 1987         |

| Row   | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|-------|---------|---------|----------|---------------|------------|
| 1     | 98022   | 47.2638 | -121.906 | 1160          | 42882      |
⋮
| 21612 | 98004   | 47.65   | -122.214 | 3930          | 25449      |
| 21613 | 98102   | 47.6298 | -122.323 | 3940          | 8800       |

### Order the dataset in decreasing order of house prices

In [16]:
sorted_df_2 = sort(df,cols = :price, rev = true)

21613x21 DataFrame
| Row   | id         | date              | price   | bedrooms | bathrooms |
|-------|------------|-------------------|---------|----------|-----------|
| 1     | 6762700020 | "20141013T000000" | 7700000 | 6        | 8.0       |
⋮
| 21612 | 40000362   | "20140506T000000" | 78000   | 2        | 1.0       |
| 21613 | 3421079032 | "20150217T000000" | 75000   | 1        | 0.0       |

| Row   | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-------|-------------|----------|--------|------------|------|-----------|
| 1     | 12050       | 27600    | 2.5    | 0          | 3    | 4         |
⋮
| 21612 | 780         | 16344    | 1.0    | 0          | 0    | 1         |
| 21613 | 670         | 43377    | 1.0    | 0          | 0    | 3         |

| Row   | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|-------|-------|------------|---------------|----------|--------------|
| 1     | 13    | 8570       | 3480          | 1910     | 1987         |
⋮
| 21612 | 5     | 780        | 0             | 1942     | 0            |
| 21613 | 3     | 670        | 0             | 1966     | 0            |

| Row   | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|-------|---------|---------|----------|---------------|------------|
| 1     | 98102   | 47.6298 | -122.323 | 3940          | 8800       |
⋮
| 21612 | 98168   | 47.4739 | -122.28  | 1700          | 10387      |
| 21613 | 98022   | 47.2638 | -121.906 | 1160          | 42882      |

## Multi level sorting

### Sort house data according to the increasing order of bedrooms and price

In [17]:
Dual_levelsort = sort(df,cols = [:bedrooms,:price])

21613x21 DataFrame
| Row   | id         | date              | price  | bedrooms | bathrooms |
|-------|------------|-------------------|--------|----------|-----------|
| 1     | 9543000205 | "20150413T000000" | 139950 | 0        | 0.0       |
⋮
| 21612 | 1773100755 | "20140821T000000" | 520000 | 11       | 3.0       |
| 21613 | 2402100895 | "20140625T000000" | 640000 | 33       | 1.75      |

| Row   | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-------|-------------|----------|--------|------------|------|-----------|
| 1     | 844         | 4269     | 1.0    | 0          | 0    | 4         |
⋮
| 21612 | 3000        | 4960     | 2.0    | 0          | 0    | 3         |
| 21613 | 1620        | 6000     | 1.0    | 0          | 0    | 5         |

| Row   | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|-------|-------|------------|---------------|----------|--------------|
| 1     | 7     | 844        | 0             | 1913     | 0            |
⋮
| 21612 | 7     | 2400       | 600           | 1918     | 1999         |
| 21613 | 7     | 1040       | 580           | 1947     | 0            |

| Row   | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|-------|---------|---------|----------|---------------|------------|
| 1     | 98001   | 47.2781 | -122.25  | 1380          | 9600       |
⋮
| 21612 | 98106   | 47.556  | -122.363 | 1420          | 4960       |
| 21613 | 98103   | 47.6878 | -122.331 | 1330          | 4700       |

### Sort data in decreasing order of bedroom and increasing order of price

In [18]:
dual_sort_2 = sort(df,cols = [:bedrooms,:price],rev = [true,false])

21613x21 DataFrame
| Row   | id         | date              | price   | bedrooms | bathrooms |
|-------|------------|-------------------|---------|----------|-----------|
| 1     | 2402100895 | "20140625T000000" | 640000  | 33       | 1.75      |
⋮
| 21612 | 6306400140 | "20140612T000000" | 1095000 | 0        | 0.0       |
| 21613 | 2954400190 | "20140624T000000" | 1295648 | 0        | 0.0       |

| Row   | sqft_living | sqft_lot | floors | waterfront | view | condition |
|-------|-------------|----------|--------|------------|------|-----------|
| 1     | 1620        | 6000     | 1.0    | 0          | 0    | 5         |
⋮
| 21612 | 3064        | 4764     | 3.5    | 0          | 2    | 3         |
| 21613 | 4810        | 28008    | 2.0    | 0          | 0    | 3         |

| Row   | grade | sqft_above | sqft_basement | yr_built | yr_renovated |
|-------|-------|------------|---------------|----------|--------------|
| 1     | 7     | 1040       | 580           | 1947     | 0            |
⋮
| 21612 | 7     | 3064       | 0             | 1990     | 0            |
| 21613 | 12    | 4810       | 0             | 1990     | 0            |

| Row   | zipcode | lat     | long     | sqft_living15 | sqft_lot15 |
|-------|---------|---------|----------|---------------|------------|
| 1     | 98103   | 47.6878 | -122.331 | 1330          | 4700       |
⋮
| 21612 | 98102   | 47.6362 | -122.322 | 2360          | 4000       |
| 21613 | 98053   | 47.6642 | -122.069 | 4740          | 35061      |

### Print the total number of houses available in each zipcode

In [25]:
Size_by_Zip = by(df,:zipcode,size);
head(Size_by_Zip)

6x2 DataFrame
| Row | zipcode | x1       |
|-----|---------|----------|
| 1   | 98001   | (362,20) |
⋮
| 5   | 98005   | (168,20) |
| 6   | 98006   | (498,20) |

Note: The DataFrames package supports the Split-Apply-Combine strategy through the by function, which takes in three arguments: (1) a DataFrame, (2) a column to split the DataFrame on, and (3) a function or expression to apply to each subset of the DataFrame.

### Print maximum price of house in all zipcode

In [23]:
maxPrice_Zipcode = by(df,:zipcode, testdf -> testdf[indmax(testdf[:price]),[:price]]);
head(maxPrice_Zipcode)

6x2 DataFrame
| Row | zipcode | price   |
|-----|---------|---------|
| 1   | 98001   | 850000  |
⋮
| 5   | 98005   | 1960000 |
| 6   | 98006   | 4208000 |

### Print minimun price of house in all zipcode

In [24]:
minPrice_Zipcode = by(df,:zipcode, testdf -> testdf[indmin(testdf[:price]),[:price]]);
head(minPrice_Zipcode)

6x2 DataFrame
| Row | zipcode | price  |
|-----|---------|--------|
| 1   | 98001   | 100000 |
⋮
| 5   | 98005   | 400000 |
| 6   | 98006   | 247500 |